### PR TITLE
Remove site selection for payment method management in me/purchases

### PR DESCRIPTION
**Pull request by jboland88**
Related to #75927

## Proposed Changes

* Adds a way to add or change payment methods for a siteless purchase (including Akismet)

## Testing Instructions

* Purchase a product from Akismet.com
* Navigate to /me/purchases and find your purchase in the list - click on it to go to the purchase management screen
* From the purchase management screen, you should now have the button/ action to change your payment method

Screenshot 2023-04-18 at 6 02 10 PM

* Test Jetpack and WordPress.com purchases as well from /me/purchases to make sure managing payment methods works as expected
* Test that payment management works as expected when accessed from the /my-sites/ routes (managing a purchase with a specific site selected in the site selector)

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Review by JessBoctor — COMMENTED** [review]
It looks good to me. I was able to edit the payment method with this PR applied: 

GH75929-AkismetPaymentMethods-Screenshot 2023-04-18 at 3 30 15 PM

I also checked for Jetpack and WordPress.com purchase management and they worked as I expected.

The only question I had about the code was if removing `siteSelected` would be a problem for renewals, or which payments are listed under each payment method. Since renewal information is stored in the subscription tables and there is no screen to see which subscriptions are listed with the payment method, I think this is all good. 

Thanks for the quick response to this issue!

**Comment by DavidRothstein** [comment]
Thanks for working on this!

It worked fine for me when switching to an **existing** payment method, but when trying to switch to a new credit card on that screen I just got "User or Token does not have access to specified site" which comes from the `upgrades/[subscription-id]/update-credit-card` endpoint.  Can anyone else reproduce that (using cards from https://stripe.com/docs/testing in store sandbox)?

As far as I can see, that error only happens if the user does not have access to the site on which the subscription exists.  But why would the user not have access to their own Akismet Siteless Checkout holding site?

**Comment by IanRamosC** [comment]
> Can anyone else reproduce that (using cards from https://stripe.com/docs/ in store sandbox)?

It worked fine for me both when changing to an existing payment method and changing to a new one - it showed the success message "Your payment method has been set."

I also tried cancelling all my active Akismet subscriptions, so the next purchase would trigger the creation of a new holding site, and it also worked for the new purchase (with the new holding site).

@DavidRothstein is there anything else that could be triggering that message to you?

**Review by IanRamosC — APPROVED** [review]
LGTM! Let's see what's causing what was mentioned in the other comments and merge it if it's ready.

**Comment by jboland88** [comment]
@DavidRothstein @IanRamosC - I was able to reproduce the issue mentioned [here](https://github.com/Automattic/wp-calypso/pull/75929#issuecomment-1514096784) where adding a new card was declined. I could reproduce this with a test user.

I have a fix for that in D108321-code which will need to get deployed ahead of this PR. Could you retest with this diff applied?

File: client/me/purchases/manage-purchase/change-payment-method/index.tsx [hunk 1]
```diff
@@ -46,14 +45,13 @@ function ChangePaymentMethod( {
...
 	const payment = useSelector( ( state ) => getByPurchaseId( state, purchaseId )?.payment );
-	const selectedSite = useSelector( getSelectedSite );
 	const { isLoading: isLoadingStoredCards } = useStoredPaymentMethods( { type: 'card' } );
 
 	const { isStripeLoading } = useStripe();
 
 	const isDataLoading =
 		! hasLoadedSites || ! hasLoadedUserPurchases || isLoadingStoredCards || isStripeLoading;
-	const isDataValid = purchase && selectedSite;
+	const isDataValid = purchase;
```
Comment (CodeyGuyDylan) [inline client/me/purchases/manage-purchase/change-payment-method/index.tsx]: This variable seems kinda irrelevant now. You could just use `! purchase` below and it should do the same thing, and likewise anywhere else this variable is used 😄

**Review by CodeyGuyDylan — APPROVED** [review]
Changes look good! Made one note about an unneeded variable but the test plan worked well as well as changing the payment card with a non-a11n user account

**Comment by DavidRothstein** [comment]
>I have a fix for that in D108321-code which will need to get deployed ahead of this PR. Could you retest with this diff applied?

I tested and confirmed it seems to be working well with that -- thanks!

File: client/me/purchases/manage-purchase/change-payment-method/index.tsx (hunk 1 above)
Comment (jboland88) [reply client/me/purchases/manage-purchase/change-payment-method/index.tsx]: Good point - had to fix another bit of logic, so updated this as well.

**Review by elliottprogrammer — APPROVED** [review]
Tested this with D108321-code applied. Tested changing & adding payment method for Akismet purchase and works great, as described. Tested changing payment method for Jetpack and WordPress.com purchases too. Can successfully turn auto-renew ON and OFF.
Everything works well for me. Code looks good, I don't see any issues. Unit tests passing too.
LGTM! 👍
